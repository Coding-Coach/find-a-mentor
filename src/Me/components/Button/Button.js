// @ts-check
import React from 'react';
import styled from 'styled-components';
import { desktop } from '../../styles/shared/devices';
import PropTypes from 'prop-types';

/**
 * @typedef {('primary' | 'secondary' | 'danger')} Skin
 * @typedef {Pick<React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>, 'onClick'>} ButtonProps
 */

const StyledButton = styled.button`
  box-sizing: border-box;
  height: 30px;
  width: 285px;
  border-radius: 3px;
  font-family: Lato;
  font-size: 14px;
  line-height: 17px;
  text-align: center;
  margin-bottom: 1rem;
  cursor: pointer;

  @media ${desktop} {
    height: 30px;
    width: 151px;
    margin: 1rem;
  }
`;

const PrimaryButton = styled(StyledButton)`
  background-color: #69d5b1;
  color: #fff;
`;

const SecondaryButton = styled(StyledButton)`
  background-color: #fff;
  border: 2px solid;
  color: #69d5b1;
`;

const DangerButton = styled(StyledButton)`
  background-color: #d56969;
  color: #fff;
`;

/**
 * @param {Skin} skin
 */
const getComponentBySkin = skin => {
  switch (skin) {
    case 'primary':
    default:
      return PrimaryButton;
    case 'secondary':
      return SecondaryButton;
    case 'danger':
      return DangerButton;
  }
};

/**
 * @param {{
 *  skin: Skin,
 * } & ButtonProps
 * } params
 */
export const Button = ({ skin = 'primary', ...props }) => {
  const ThemedButton = getComponentBySkin(skin);
  return <ThemedButton {...props} />;
};

Button.propTypes = {
  skin: PropTypes.oneOf(['primary', 'secondary', 'danger']),
};
