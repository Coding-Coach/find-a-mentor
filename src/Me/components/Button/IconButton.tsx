import styled from 'styled-components/macro';
import classNames from 'classnames';

type IconButtonProps = {
  icon: string;
  onClick: () => void;
  buttonProps?: React.ButtonHTMLAttributes<HTMLButtonElement>;
  color?: string;
  size?: 'sm' | 'lg' | '2x' | '3x';
};

const Button = styled.button`
  background: none;
  cursor: pointer;
`;

export const IconButton = ({ onClick, icon, size, color, buttonProps }: IconButtonProps) => {
  return (
    <Button onClick={onClick} {...buttonProps}>
      {typeof icon === 'string' && (
        <i
          style={{ color }}
          className={classNames(['fa', `fa-${icon}`, { [`fa-${size}`]: size }])}
        />
      )}
    </Button>
  );
};
