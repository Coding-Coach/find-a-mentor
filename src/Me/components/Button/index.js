export { Button as default } from './Button';
