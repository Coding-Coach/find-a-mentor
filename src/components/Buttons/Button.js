import React from 'react';
import classNames from 'classnames';
import styles from './Button.scss';
import PropTypes from 'prop-types';
/**
 *
 * @param {children} props
 * @param {type} props  primary, danger, secondary
 * @param {size} props  small, medium, large,
 * @param {disabled} props  true, false
 * @param {fullWidth} props Button width is 100% of parent
 * @param {typography} props capitalize, uppercase, lowercase, only
 * @param {floating} props Adds a shadow under the button
 */
const Button = (props) => {
  const {
    children,
    floating,
    type,
    size,
    shape,
    typography,
    disabled,
    fullWidth,
    id,
    onClick,
  } = props;

  return (
    <button
      id={id}
      disabled={disabled}
      onClick={onClick}
      className={classNames({
        button: true,
        [type]: !disabled,
        disabled: disabled,
        'shadow-floating hover:shadow': floating,
        'w-full': fullWidth,
        [size]: true,
        [shape]: true,
        [typography]: true,
      })}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  type: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'disabled', 'danger']),
  typography: PropTypes.oneOf(['uppercase', 'capitalize', 'lowercase', 'none']),
  shape: PropTypes.oneOf(['square', 'rounded', 'pill']),
  children: PropTypes.node,
  disabled: PropTypes.bool,
  floating: PropTypes.bool,
  fullWidth: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  id: PropTypes.string,
};

Button.defaultProps = {
  size: 'medium',
  type: 'primary',
  typography: 'uppercase',
  shape: 'rounded',
  floating: false,
};

export default Button;