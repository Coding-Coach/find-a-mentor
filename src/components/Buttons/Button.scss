@import '../../theme/base/variables';
@import '../../theme/base/mediaqueries';
@import '../../theme/base/utils';
@import '../../theme/base/spacing';

.button {
  display: inline-block;
  min-height: $button-default;
  line-height: inherit;
  min-width: 96px;
  box-sizing: border-box;
  text-transform: uppercase;
  text-align: center;
  white-space: nowrap;
  outline: none;
  border: none;
  border-radius: 3px;
  transition-property: all;
  transition-duration: 0.15s;
  font-size: $font-xs;

  &.primary {
    background: $color-primary;
    box-shadow: 0px 0px 0px 0px inset $color-primary-darker,
      0px 0px 20px 3px $color-primary-blur;
    color: $color-white;

    &:hover {
      box-shadow: 0px -5px 0px 0px inset $color-primary-darker,
        0px 0px 20px 3px $color-primary-blur;
    }

    &:active {
      background: $color-primary-darker;
    }
  }

  &.secondary {
    background: transparent;
    box-shadow: 0px 0px 0px 0px inset $color-primary,
    0px 0px 20px 0px inset $color-primary-blur;
    color: $color-primary;
    box-sizing: border-box;
    cursor: pointer;
    border: 2px solid $color-primary;

    &:hover {
      box-shadow: 0px -3px 0px 0px inset $color-primary,
        0px 0px 20px 0px inset $color-primary-blur;
    }

    &:active {
      background: $color-primary;
      box-shadow: none;
      color: $color-white;
    }
  }

  &.tertiary {
    background: transparent;
    color: $color-primary-darker;
    box-sizing: border-box;
    cursor: pointer;

    &:hover {
      text-decoration: underline;
    }
  }

  &.danger {
    background: $color-danger;
    box-shadow: 0px 0px 0px 0px inset $color-danger-darker,
    0px 0px 20px 3px $color-danger-blur;
    color: $color-white;

    &:hover {
      box-shadow: 0px -5px 0px 0px inset $color-danger-darker,
        0px 0px 20px 3px $color-danger-blur;
    }

    &:active {
      background: $color-danger-darker;
    }
  }

  &:disabled {
    background: $color-grey-light;
    box-shadow: 0px 0px 20px 3px $color-grey-light;
    color: $color-grey;

    &:hover, &:active {
      transform: none;
      box-shadow: 0px 0px 20px 3px $color-grey-light;
      background: $color-grey-light;
      cursor: not-allowed;
    }
  }

  &.transparent-white {
    background: transparent;
    color: $color-white;
    box-sizing: border-box;
    cursor: pointer;

    &:hover {
      text-decoration: underline;
    }
  }

  /* Common Hover and Active styles for all Buttons*/
  &:hover {
    cursor: pointer;
  }

  &:active {
    text-shadow: none;
  }


  &.default {
    padding: 0px $font-s;
  }

  &.large {
    padding: 0px $font-l;
    min-height: $button-large;
    line-height: $button-large;
    font-size: $font-l;
  }

  &.small {
    padding: 0px $font-xs;
    min-height: $button-small;
    line-height: $button-small;
    font-size: $font-xs
  }

  @include respond-to(mobile) {
    &.small {
      padding: 0px $font-xxs;
      font-size: $font-xxs;
    }
  }

  &.fullWidth {
    width: 100%;
  }

  // For icons and svgs
  i, svg {
    vertical-align: middle;
    margin-right: 4px;
  }
}