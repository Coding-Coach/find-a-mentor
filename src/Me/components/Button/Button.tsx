import React, { FC } from 'react';
import styled from 'styled-components';
import { Loader } from '../../../components/Loader';

type Skin = 'primary' | 'secondary' | 'danger';
type ButtonProps = Pick<
  React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  >,
  'onClick' | 'id' | 'disabled' | 'type' | 'name' | 'children' | 'title'
> & {
  skin?: Skin;
  isLoading?: boolean;
};

const StyledButton = styled.button`
  box-sizing: border-box;
  height: 30px;
  border-radius: 3px;
  font-family: Lato, sans-serif;
  font-size: 14px;
  line-height: 17px;
  text-align: center;
  cursor: pointer;

  &:disabled {
    opacity: 0.5;
  }
`;

const PrimaryButton = styled(StyledButton)`
  background-color: #69d5b1;
  color: #fff;
`;

const SecondaryButton = styled(StyledButton)`
  background-color: #fff;
  border: 2px solid;
  color: #69d5b1;
`;

const DangerButton = styled(StyledButton)`
  background-color: #d56969;
  color: #fff;
`;

const getComponentBySkin = (skin: Skin) => {
  switch (skin) {
    case 'primary':
    default:
      return PrimaryButton;
    case 'secondary':
      return SecondaryButton;
    case 'danger':
      return DangerButton;
  }
};

export const Button: FC<ButtonProps> = ({
  skin = 'primary',
  isLoading = false,
  children,
  ...props
}) => {
  const ThemedButton = getComponentBySkin(skin);
  return (
    <ThemedButton {...props}>{isLoading ? <Loader /> : children}</ThemedButton>
  );
};
