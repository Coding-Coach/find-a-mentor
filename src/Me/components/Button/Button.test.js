import { render } from '@testing-library/react';
import React from 'react';
import Button from './';

describe('Button component', () => {
  it('renders primary button by default', () => {
    const { container } = render(
      <Button onClick={() => {}}>primary button</Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders secondary button when specified with skin prop', () => {
    const { container } = render(
      <Button skin="secondary" onClick={() => {}}>
        secondary button
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders danger button when specified with skin prop', () => {
    const { container } = render(
      <Button skin="danger" onClick={() => {}}>
        danger button
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('passes down other props to the button tag', () => {
    const { getByTestId } = render(
      <Button data-testid="test-button" onClick={() => {}}>
        click me
      </Button>
    );
    expect(getByTestId('test-button')).toBeInTheDocument();
  });
});
