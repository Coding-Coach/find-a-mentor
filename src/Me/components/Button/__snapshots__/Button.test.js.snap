// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders danger button when specified with skin prop 1`] = `
.c0 {
  box-sizing: border-box;
  height: 30px;
  width: 285px;
  border-radius: 3px;
  font-family: Lato;
  font-size: 14px;
  line-height: 17px;
  text-align: center;
  margin-bottom: 1rem;
  cursor: pointer;
  background-color: #d56969;
  color: #fff;
}

@media (min-width:801px) {
  .c0 {
    height: 30px;
    width: 151px;
    margin: 1rem;
  }
}

<div>
  <button
    class="sc-AxjAm c0"
  >
    danger button
  </button>
</div>
`;

exports[`Button component renders primary button by default 1`] = `
.c0 {
  box-sizing: border-box;
  height: 30px;
  width: 285px;
  border-radius: 3px;
  font-family: Lato;
  font-size: 14px;
  line-height: 17px;
  text-align: center;
  margin-bottom: 1rem;
  cursor: pointer;
  background-color: #69d5b1;
  color: #fff;
}

@media (min-width:801px) {
  .c0 {
    height: 30px;
    width: 151px;
    margin: 1rem;
  }
}

<button
  class="sc-AxjAm c0"
>
  primary button
</button>
`;

exports[`Button component renders secondary button when specified with skin prop 1`] = `
.c0 {
  box-sizing: border-box;
  height: 30px;
  width: 285px;
  border-radius: 3px;
  font-family: Lato;
  font-size: 14px;
  line-height: 17px;
  text-align: center;
  margin-bottom: 1rem;
  cursor: pointer;
  background-color: #fff;
  border: 2px solid;
  color: #69d5b1;
}

@media (min-width:801px) {
  .c0 {
    height: 30px;
    width: 151px;
    margin: 1rem;
  }
}

<div>
  <button
    class="sc-AxjAm c0"
  >
    secondary button
  </button>
</div>
`;
